Leave coding page patch

AstridKery opened the pull request:
This branch patches the leave coding page issues we were seeing where users were locked out of the admin page. It should also give back assigned data. 

NOTE: here leaving the coding page means un-assigning all data assigned to that user, and removing them from the admin-progress table which locks certain admin tables from other admin's use

The changes are as follows:

* when someone enters the coding page for one project, it leaves the coding page for that user for all other projects they have access to
* when someone logs out it leaves the coding page for all projects they have access to

I'd appreciate it if you could both test this thoroughly and let me know what you think! Thanks!

AstridKery commented:
Are they able to edit the adjudicate/skipped table at the same time? I think that's what they get locked out of, not the admin page.

dsteedRTI commented:
Unfortunately I'm still seeing the `Another admin is currently using this page. Please check back later.`

Steps to reproduce:

- Chrome, logged in as root user, add test_user as an Admin to the Three User Project
- Safari, logged in as test_user, goes to Annotate Tab and then over to Skipped Cards
- Chrome (root) goes to Skipped Cards, sees the `Another admin is currently using this page. Please check back later.` but this is expected...
- Safari (test_user) leave Skipped Cards and go to projects listing
- Chrome (root) refreshes, annotates a card, then goes to Skipped Cards and still sees `Another admin is currently using this page. Please check back later.`

However, do **note** that when I log out of test_user on Safari and then try root's Chrome window, now it works as expected and root can access the Skipped Cards tab.

AstridKery commented:
Right! So this behavior does actually fall in line with what I expected (although I'll agree that it's not ideal). With this new change, admin A must either go to a different project annotate page, or log out for the project to unlock for other admin. In the past they could just leave the annotate page, but this function was disabled by chrome so we had to find alternatives. @dsteedRTI do you think going to the project list page or details page should also be an unlock condition? That might solve some of the frustration.

I do want to note one thing though. The annotate page which includes the admin skipped tab is all one piece, and access to it has always been considered one thing. Historically, if admin A goes into project 1 and is annotating cards, admin B **will** be locked out of the skipped tab until admin A leaves the annotate page. It's never been possible for admin B to see the skip table while another admin was labeling. Not saying that's not a weird design choice, but it's not a new one.

peridoteagle commented:
Thanks for working on this, Caroline! Love what I'm seeing. I can replicate that if a user logs out on Chrome or Firefox, the entry in the adminprogress table goes away. I can also confirm that the entry goes away and a new entry is created in the adminprogress table when the person switches projects.

I agree with Durk - I think adding the same functionality for if a user returns to the project list page would be ideal. I'm not sure going to the Details page should be as high priority.

Two user interface type updates:
1. It would likely be helpful to change the error message of "Another admin is currently using this page. Please check back later." to something more informative like "Another admin is currently using this page. This page will become available when the admin returns to the project list page, changes projects, or logs out"
2. I've got a project with ~30 unlabeled items; the second user to login sees an error that says "No More Cards" but I think the more accurate error would be "All Cards Are Assigned". Does this current PR also still impact the card assignment?

Finally one item to think about - what if there was someone who was in the middle of adjudicating, closed their window without logging out or changing projects, and then quit RTI and had no ability to access their account? Do we have a way to force their logout or similar?

AstridKery commented:
Ok I've added updates so that they should leave the coding page if they go to details or project list. I also updated the messages for the admin lockout and out of cards. @peridoteagle and @dsteedRTI feel free to take another look on Monday, thanks!

AstridKery commented:
One question for @peridoteagle : currently they don't get their assignments  or lock removed if they go to the admin page. Should they? It's technically a view page like details and they can't really affect the project there, but I could see someone being confused that the project got given to another admin because they went there.

peridoteagle commented on frontend/src/components/Smart/index.jsx:
```diff
@@ -35,7 +35,7 @@ class Smart extends React.Component {
                 <Tab eventKey={3} transition={false} title="Fix Skew" className="full card">
                     <div className="cardContent">
-                        <h2>Another admin is currently using this page. Please check back later.</h2>
+                        <h2>Another admin is currently using this page. This page will become available when the admin returns to the project list page, changes projects, or logs out.</h2>
```
Agree with Durk's comment. Would also include something like "this page will become available **upon refresh** ..." or some similar instruction to refresh. I am imagining a scenario where someone gets this error message, slacks the team to have admins change projects, and then doesn't see anything update (because they need to manually refresh the page)

Could consider changing "Another admin" to a username if it's straightforward, but I'm not particularly attached to this feature unless someone requests it

AstridKery replied:
honestly it's a decent amount of work to add it, so maybe we could put it as an issue to get this merged in?

AstridKery replied:
also updated admin locked message to include details page

peridoteagle commented:
Wow, this is fantastic! Tested on the staging server and works SO great!
